backend.go

```diff
@@ -87,81 +94,223 @@ func (b *azureSecretBackend) periodicFunc(ctx context.Context, sys *logical.Requ
+
+	err = b.saveConfig(ctx, config, storage)
+	if err != nil {
+		return err
+	}
+
+	b.updatePassword = false
+
+	return nil
+}
 
-		apps, err := client.provider.ListApplications(ctx, fmt.Sprintf("appId eq '%s'", config.ClientID))
+func (b *azureSecretBackend) refreshSPCredentials(ctx context.Context, storage logical.Storage) error {
+	roleNames, err := storage.List(ctx, rolesStoragePath+"/")
+	if err != nil {
+		return fmt.Errorf("error listing roles: %w", err)
+	}
+
+	b.Logger().Debug("listed roles", "roles", roleNames)
+
+	expirationCutoff := time.Now().Add(spCredsRefreshBuffer)
+	for _, roleName := range roleNames {
+		err = b.refreshSPCredentialsForRole(ctx, storage, roleName, expirationCutoff)
 		if err != nil {
-			return err
+			b.Logger().Error("failed to refresh service principal credentials for role", "role", roleName, "error", err)
 		}
+	}
 
-		if len(apps) == 0 {
-			return fmt.Errorf("no application found")
-		}
-		if len(apps) > 1 {
-			return fmt.Errorf("multiple applications found - double check your client_id")
-		}
+	return nil
+}
 
-		app := apps[0]
+func (b *azureSecretBackend) refreshSPCredentialsForRole(ctx context.Context, storage logical.Storage, roleName string, expirationCutoff time.Time) error {
+	lock := locksutil.LockForKey(b.appLocks, roleName)
+	lock.Lock()
+	defer lock.Unlock()
 
-		credsToDelete := []string{}
-		for _, cred := range app.PasswordCredentials {
-			if *cred.KeyID != config.NewClientSecretKeyID {
-				credsToDelete = append(credsToDelete, *cred.KeyID)
-			}
-		}
+	client, err := b.getClient(ctx, storage)
+	if err != nil {
+		return err
+	}
 
-		if len(credsToDelete) != 0 {
-			b.Logger().Debug("periodic func", "rotate-root", "removing old passwords from Azure")
-			err = removeApplicationPasswords(ctx, client.provider, *app.ID, credsToDelete...)
-			if err != nil {
-				return err
-			}
+	b.Logger().Debug("checking service principal credentials", "role", roleName)
+
+	role, err := getRole(ctx, roleName, storage)
+	if err != nil {
+		return fmt.Errorf("failed to get role %q: %w", roleName, err)
+	}
+
+	if role.ApplicationType != applicationTypeDynamic {
+		return nil
+	} else if role.Credentials == nil {
+		// This should be impossible
+		return fmt.Errorf("dynamic application role missing credentials. role: %s", roleName)
+	}
+
+	if role.Credentials.ExpiresAt.After(expirationCutoff) {
+		return nil
+	}
+
+	b.Logger().Debug("retrieving service principal", "id", role.ServicePrincipalID)
+	servicePrincipal, err := client.getSP(ctx, role.ServicePrincipalID)
+	if err != nil {
+		return fmt.Errorf("error getting service principal: %w", err)
+	}
+
+	// we don't have the keyID and the service principal has a single password that hasn't expired, so we assume that is the
+	// password we are currently using and update the role.
+	if role.Credentials.KeyId == "" && len(servicePrincipal.PasswordCredentials) == 1 && servicePrincipal.PasswordCredentials[0].EndDate.After(expirationCutoff) {
```

Comment:
is it worth it to just iterate over `servicePrincipal.PasswordCredentials`?

Is it ever wrong to refresh those credentials?

Reply:
This check is only needed for importing the key ID and expiration time for roles created by the previous plugin version where the credentials haven't expired yet. Those roles could have a single un-expired credential or a single expired credential.